src/translations.h

```diff
@@ -50,15 +59,45 @@
 #  define ATTRIBUTE_FORMAT_ARG(a)
 #endif
 
-const char *_( const char *msg ) ATTRIBUTE_FORMAT_ARG( 1 );
-inline const char *_( const char *msg )
+namespace details
+{
+// same as _(), but without local cache
+const char *_translate_internal( const char *msg ) ATTRIBUTE_FORMAT_ARG( 1 );
+
+inline const char *_translate_internal( const char *msg )
 {
     return msg[0] == '\0' ? msg : gettext( msg );
 }
-inline std::string _( const std::string &msg )
+
+// same as _(), but without local cache
+inline std::string _translate_internal( const std::string &msg )
 {
-    return _( msg.c_str() );
+    return _translate_internal( msg.c_str() );
 }
+} // namespace details
+
+// Implementation note:
+// _cached_arg_ptr is needed to safeguard against invalid _cached_translation in the
+// scenario when `msg` is `char *` that doesn't have translation
+// (in this case `_cached_translation` becomes equal to `msg`, according to `gettext` specs).
+// If original `msg` later is invalidated, then `_cached_translation` consequently becomes invalid.
+// Comparing address of the argument will catch that situation and ensure that invalid `_cached_translation`
+// is never returned.
+#define _( msg ) \
+    ( ( []( const auto & arg )-> const auto & { \
+        static int _cached_lang_version = details::get_current_language_version(); \
+        static auto * _cached_arg_ptr = &arg; \
```

Comment:
If `msg` is `const char *`, the inferred type of `arg` is `const char *const &`, so this cached pointer becomes the stack address of the pointer instead of the pointer itself. If `arg` is then invalidated and points to the same string at a different memory location and `_( msg )` is called again, the invalid `_cached_translation` will probably still get returned.

I suggest either doing all the operations with `std::string` to avoid the issue altogether, or adding some unit test to ensure this macro is acting correctly.

Also, if `msg` is `std::string` and is at a different address when this macro is called again (when `msg` is on the heap, for example), comparing the pointer would completely disable the cache and degrade the performance.

Reply:
Thanks for the review.
You are right regarding the stack address, although I think `_cached_translation` won't be returned in the described scenario, I think there might be safer ways to implement this.